---
title: ""
always_allow_html: yes
output:
  rmarkdown::html_document:
        toc: false
        toc_depth: 2
        number_sections: false
        theme: readable
        highlight: default
        toc_float: false
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(results="asis")
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(echo=FALSE)
knitr::opts_chunk$set(warning=FALSE)
knitr::opts_chunk$set(message=FALSE)
knitr::opts_chunk$set(fig_cap="yes")
```

```{r, eval=FALSE, echo=FALSE}
cd ~/Sites/XGR/PiER-site
################################################
R
library(tidyverse)
## rmarkdown::render + bookdown::html_document2
rmarkdown::render("now.RMD_eV2CG.Rmd", bookdown::html_document2(number_sections=F,theme="readable", hightlight="default"), output_dir="/Users/hfang/Sites/XGR/PiER-site/app/examples")

rmarkdown::render("pier_app/public/RMD_eV2CG.Rmd", bookdown::html_document2(number_sections=F,theme="readable", hightlight="default"))

/home/hfang/R-3.6.2/bin/R
rmarkdown::pandoc_available()
Sys.setenv(RSTUDIO_PANDOC="/home/hfang/.local/bin")
rmarkdown::render("pier_app/public/RMD_eV2CG.Rmd", bookdown::html_document2(number_sections=F,theme="readable", hightlight="default"))



q('no')
################################################
```


# {.tabset .tabset-fade}


## **Input into eV2CG**
Completed at `r format(Sys.time(), "%a %b %d %X %Y")` (`r Sys.timezone()`), with the runtime of `r ls_rmd$runtime` on the server side (not including the runtime on the client/user side).

> **An interactive table** of user-input SNPs, with 1st column `SNPs` for dbSNP rsIDs, 2nd column `P-values` for significance info.

```{r}
data_input <- ls_rmd$data_input %>% transmute(SNPs=SNP, `P-values`=PVALUE)

message(sprintf("'%d' '%s' (%s) ...", nrow(data_input), getwd(), as.character(Sys.time())), appendLF=TRUE)

data_input %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```


## **Output: core genes**

> **Manhattan plot** illustrates scored core genes color-coded by chromosomes, with the top `r ls_rmd$highlight_top` genes labelled. Also provided is an editable PDF file for download.

```{r}
hlink_pdf <- oAuxEmbed(ls_rmd$pdf_manhattan, link.text="Download PDF")
```
<div class="text-right">`r hlink_pdf`</div>

```{r manhattan, fig.cap="Manhattan plot illustrating scores (y-axis) for core genes (color-coded by chromosomes; x-axis), with the top scored genes labelled.", fig.width=9, fig.height=3, echo=FALSE, eval=T, fig.align="center",out.width="100%"}
library(png)
library(grid)
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'dTarget_manhattan.png')
filename <- ls_rmd$png_manhattan
img <- readPNG(filename)
grid.raster(img)
```

<hr>

> **An interactive table** of core genes linked from the input SNPs, with the column `Scores` quantifying the degree of genes responsible for genetic associations (capped at 100). `Core genes` are cross-referenced and hyperlinked to [GeneCards](https://www.genecards.org). Also provided is the column `Evidence` used to define core genes; please refer to **`Evidence table`** for details.

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG.xlsx')
filename <- ls_rmd$xlsx_CG

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% mutate(`Core genes`=sprintf("<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene=%s' target='%s''>%s</a>", Gene, Gene, Gene)) %>% transmute(`Core genes`, Scores=signif(Scores,digits=4), Evidence, Description)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```

> **Evidence table** for core genes, showing which SNPs (see the column `SNPs`) are used to define core genes (the column `Core genes`) based on which evidence (see the column `Evidence`). The column `SNP type` tells the SNP type (either `Input` for use-input SNPs or `LD` for LD SNPs). Notably, the column `Evidence` details datasets used: the prefix `Proximity_` indicative of SNPs in the proximity, the prefix `PCHiC_` for PCHi-C datasets, and the prefix `QTL_` for e/pQTL datasets.

```{r}
#filename <- str_c('/Users/hfang/Sites/XGR/PiER-site/app/examples/', 'eV2CG_evidence.xlsx')
filename <- ls_rmd$xlsx_CG_evidence

message(sprintf("'%s' '%s' (%s) ...", filename, getwd(), as.character(Sys.time())), appendLF=TRUE)

tb <- openxlsx::read.xlsx(filename) %>% as_tibble() %>% transmute(`Core genes`=Gene, SNPs=SNP, `SNP type`=SNP_type, Evidence)

tb %>% DT::datatable(options=list(pageLength=5,searchHighlight=T,dom='Bfrtip',buttons=c('csv','copy'), dom='t',scrollX=T), style='default', caption="", rownames=FALSE, escape=F, extensions=c('Buttons'))
```

